[WIP][BEAM-12097] ZetaSQL OVER clause converted to Logical SQL Plan

roger-mike opened the pull request:
GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

roger-mike commented on sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/AnalyticScanConverter.java:
```diff
@@ -0,0 +1,35 @@
+
+import java.util.List;
+
+public class AnalyticScanConverter extends RelConverter<ResolvedAnalyticScan> {
```
Created this converter to handle the ```ResolvedAnalyticScan``` node. However, not sure if this is a good approach or if the ```AggregateScanConverter``` should be reused somehow. Could you give me some pointers? @apilloud @ibzib

ibzib replied:
We use AggregateScanConverter to convert ZetaSql's ResolvedAggregateScan to Calcite's LogicalAggregate. Similarly, we need AnalyticScanConverter to transform ResolvedAnalyticScan to a RelNode. I believe LogicalWindow is the type of output RelNode we want.

Because the input and output types are presumably different, I don't think it will be possible to reuse AggregateScanConverter directly. I'll admit I'm a bit confused by the TODO in AggregateScanConverter (mentioned in the JIRA).

https://github.com/apache/beam/blob/43cd356f865fa52bcce217c613e747a9220136e0/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/AggregateScanConverter.java#L149

Since analytic scans are a separate class, I'm not sure how AggregateScanConverter would handle them. Andrew probably knows more, but he's still on leave for a couple more weeks.

However, I do think we can reuse at least some of the logic in AggregateScanConverter.

https://github.com/apache/beam/blob/43cd356f865fa52bcce217c613e747a9220136e0/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/AggregateScanConverter.java#L123-L127

It looks like the input of ResolvedAnalyticScan is also a scan rather than a project, so we will probably have to do something similar in AnalyticScanConverter. Perhaps since ResolvedAnalyticScan and ResolvedAggregateScan share the ResolvedScan parent class, we can make `convertAggregateScanInputScanToLogicalProject` more generic, and share it for both cases.

roger-mike commented:
Hi @ibzib @apilloud. I have some questions about the equivalencies between the ZetaSQL AST and CalciteSQL AST.  For this query: 
```sql
select sum(Key) over (PARTITION BY Value), count(Key) over (ORDER BY Key) From KeyValue
```
We get the following AST:
![Screen Shot 2022-03-01 at 16 26 43](https://user-images.githubusercontent.com/61989986/157715464-a1c26e24-23ef-459b-a81f-858313deccf5.png)

I'm still trying to work out the conversion of the inner classes needed for ```LogicalWindow```. I think the ```ResolvedAnalyticFunctionGroup``` could be mapped to a ```Window.Group``` but still not sure how to convert ```ResolvedWindowPartitioning```, ```ResolvedWindowOrdering```, ```ResolvedAnalyticFunctionCall```.

How do you find these equivalencies? Could you give me some suggestions on how to convert this AST? Thank you.

apilloud commented:
What I normally do is write the equivalent statement in Calcite SQL and then look at the logical plan which is logged here: https://github.com/apache/beam/blob/07aaec344aef38ac7034e65c909ed75893a9d48d/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/CalciteQueryPlanner.java#L189

roger-mike commented:
Hi @apilloud, thanks for the advice. As it is now, I'm getting the following error when running `testAnalyticOver`. The implementation is still incomplete but wanted to ask you if you have an idea of why this is happening. Thanks a lot.
```
index (3) must be less than size (3)
java.lang.IndexOutOfBoundsException: index (3) must be less than size (3)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1404)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1386)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:77)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ExpressionConverter.retrieveRexNode(ExpressionConverter.java:213)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ProjectScanConverter.convert(ProjectScanConverter.java:46)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ProjectScanConverter.convert(ProjectScanConverter.java:30)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convertNode(QueryStatementConverter.java:91)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convert(QueryStatementConverter.java:78)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convertRootQuery(QueryStatementConverter.java:43)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLPlannerImpl.rel(ZetaSQLPlannerImpl.java:98)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRelInternal(ZetaSQLQueryPlanner.java:198)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRel(ZetaSQLQueryPlanner.java:186)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRel(ZetaSQLQueryPlanner.java:170)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSqlDialectSpecTest.testAnalyticOver(ZetaSqlDialectSpecTest.java:3582)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.apache.beam.sdk.testing.TestPipeline$1.evaluate(TestPipeline.java:323)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

apilloud commented:
It looks like you are failing in this line of code: https://github.com/apache/beam/blob/77e924fd5a025326b50352c29598ea1eb48d385a/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java#L213

What this means is something in a previous step was converted from ZetaSQL to Calcite wrong. There is a ZetaSQL ResolvedProjectScan that references the forth column (index 3) but the calcite List<RelDataTypeField> only has three columns.

You might be able to add some asserts earlier in the conversion to help you find the issue.

roger-mike commented:
> It looks like you are failing in this line of code:
> 
> https://github.com/apache/beam/blob/77e924fd5a025326b50352c29598ea1eb48d385a/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java#L213
> 
> What this means is something in a previous step was converted from ZetaSQL to Calcite wrong. There is a ZetaSQL ResolvedProjectScan that references the forth column (index 3) but the calcite List only has three columns.
> 
> You might be able to add some asserts earlier in the conversion to help you find the issue.

I'm currently working on the [function](https://github.com/apache/beam/pull/16911/files#diff-b497391b74a7130be5217a2d4ede920fe66059444bc248a6710ad9b6361ac2b3R81) that converts the input, I noticed the error might be occurring because [this](https://github.com/apache/beam/pull/16911/files#diff-b497391b74a7130be5217a2d4ede920fe66059444bc248a6710ad9b6361ac2b3R104) ```projects```  parameter doesn't have all the ```RexNode``` objects that it needs.

I took a look at the ```AggregateScanConverter``` implementation of this [input conversion](https://github.com/apache/beam/blob/4dddaf9c3d70b5b858c16265abee4cdbf0c72631/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/AggregateScanConverter.java#L121) and it seems that it adds  ```groupByList``` and ```aggregateList``` from the zetaNode as ```RexNodes``` to the ```LogicalProject``` as ```projects``` and ```fieldNames```. I'm not sure how to make a similiar procedure for the ```AnalyticScan```, given that it doesn't have either a ```groupByList``` or an ```aggregateList```. 

My question here is, what is the logic or purpose of converting and adding these lists (```projects``` and ```fieldNames```) to a ```LogicalProject```? @apilloud @ibzib

apilloud commented:
The question is essentially what is the purpose of the ZetaSQL translation library. `AggregateScan` is a SQL query with a group by for example `select value*2, sum(key+1) from KeyValue group by value` This would give a groupByList of [value*2] and aggregate_list of [sum(key+1)]. The LogicalProject is the first step of simplifying the data, it basicly translates to a simple ParDo. We take the input data and apply any simple transforms, so you get: LogicalProject [value*2, key+1]. The input to the logical project may have other columns that are dropped. After the logical project you are left with a simpler SQL query:
`SELECT col1, sum(col2) from LogicalProject group by col1`.

In AnalyticScan, it appears the equivalent is `functionGroupList`.

roger-mike commented:
Hi @apilloud, I asked the Calcite devs about this task, specifically about how to create a Relational Operator Tree for a window function.

I changed the approach and added the window functions as ```RexOver``` to a ```LogicalProject```, instead of building the tree from a ```LogicalWindow```.  This way we build a simpler tree and let part of the conversion to the Calcite planner. This is the comment I received from Julian Hyde on that matter:

> It’s complicated. Windowed aggregates start off as instances of RexOver expressions inside a Project and are later converted (by ProjectToWindowRule [1]) into Window relational operators.
> 
> Both of these representations exist in RelNode-land and I’m not sure which should be considered ‘canonical’. Window is a little bit more ‘physical’ than Project. At some stage in planning, it is assumed that all RexOver instances have been converted.
> 
>
>  Julian
> 
> [1] https://github.com/apache/calcite/blob/bebe473fab2e242736614659ed6e5d04eeeb8bf5/core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java#L74 <https://github.com/apache/calcite/blob/bebe473fab2e242736614659ed6e5d04eeeb8bf5/core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java#L74>

So far so good. I tested and it seems it creates a valid BEAMPlan. The issue here is that I had to hardcode the value of this [index ]([beam/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java](https://github.com/apache/beam/blob/77e924fd5a025326b50352c29598ea1eb48d385a/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java#L213))to make it point to a valid column reference. Not sure how to solve this issue. I was hoping you could give me some pointers on this. Thank you.

P.S.
I also got this one comment, I think the Calcite community might be interested in this translation feature. I said I was going to pass the message. What do you think?

> It’s very interesting that you want to convert ZetaSQL to and from Calcite. I think other people will be interested in this. (It would help some of the things I am doing in my $dayjob at Google.) Would you consider contributing this as a new module in Calcite? Or creating a standalone project on GitHub?
> 
> Julian

roger-mike commented:
R: @apilloud could you take a look? Thanks.

apilloud commented:
Unfortunately your link is broken. Is this the index? Looking at the generated code, I think hardcoding an index of `0` is reasonable as analytic functions only have one signature. https://github.com/apache/beam/pull/16911/files#diff-b497391b74a7130be5217a2d4ede920fe66059444bc248a6710ad9b6361ac2b3R105

apilloud commented:
It sounds like either `indexOfResolvedColumnInExprList` or `indexOfProjectionColumnRef` aren't producing valid indexes, so the mapping isn't working. It looks like this code maps the ZetaSQL ResolvedColumn to a Calcite column in the fieldList. I would expect this to hit the `Cannot find %s in fieldList %s` exception just above the line you pointed to.

Could you push your change with the hardcoded constant to this PR? Do you have a test failure that demonstrates this?

roger-mike commented:
> It sounds like either `indexOfResolvedColumnInExprList` or `indexOfProjectionColumnRef` aren't producing valid indexes, so the mapping isn't working. It looks like this code maps the ZetaSQL ResolvedColumn to a Calcite column in the fieldList. I would expect this to hit the `Cannot find %s in fieldList %s` exception just above the line you pointed to.
> 
> Could you push your change with the hardcoded constant to this PR? Do you have a test failure that demonstrates this?

This is the index I hardcoded:
https://github.com/apache/beam/blob/2d5b5f18e5b91f3094a323dd3785d752c90da8c1/sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java#L212

When I run ```testAnalyticOver``` with the hardcoded index it would produce the following Beam Plan:
```
BeamZetaSqlCalcRel(expr#0..2=[{inputs}], expr#3=[SUM($t0) OVER ()], $col1=[$t3])
  BeamIOSourceRel(table=[[beam, KeyValue]])
```
Otherwise, I fails like this:

```
index (3) must be less than size (1)
java.lang.IndexOutOfBoundsException: index (3) must be less than size (1)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1404)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.base.Preconditions.checkElementIndex(Preconditions.java:1386)
	at org.apache.beam.vendor.calcite.v1_28_0.com.google.common.collect.RegularImmutableList.get(RegularImmutableList.java:77)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ExpressionConverter.retrieveRexNode(ExpressionConverter.java:213)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ProjectScanConverter.convert(ProjectScanConverter.java:46)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.ProjectScanConverter.convert(ProjectScanConverter.java:30)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convertNode(QueryStatementConverter.java:91)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convert(QueryStatementConverter.java:78)
	at org.apache.beam.sdk.extensions.sql.zetasql.translation.QueryStatementConverter.convertRootQuery(QueryStatementConverter.java:43)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLPlannerImpl.rel(ZetaSQLPlannerImpl.java:98)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRelInternal(ZetaSQLQueryPlanner.java:198)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRel(ZetaSQLQueryPlanner.java:186)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSQLQueryPlanner.convertToBeamRel(ZetaSQLQueryPlanner.java:170)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSqlDialectSpecTest.testAnalyticOver(ZetaSqlDialectSpecTest.java:3580)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.apache.beam.sdk.testing.TestPipeline$1.evaluate(TestPipeline.java:323)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```
Thanks for your help @apilloud.

apilloud commented:
I printed node and fieldlist at the beginning of retrieveRexNode.
```
    Node: ProjectScan
    +-column_list=[$analytic.$analytic1#5]
    +-input_scan=
      +-AnalyticScan
        +-column_list=[KeyValue.Key#1, KeyValue.Value#2, KeyValue.ts#3, $analytic.$analytic1#5]
        +-input_scan=
        | +-TableScan(column_list=KeyValue.[Key#1, Value#2, ts#3], table=KeyValue, column_index_list=[0, 1, 2])
        +-function_group_list=
          +-AnalyticFunctionGroup
            +-analytic_function_list=
              +-$analytic1#5 :=
                +-AnalyticFunctionCall(ZetaSQL:sum(INT64) -> INT64)
                  +-ColumnRef(type=INT64, column=KeyValue.Key#1)
                  +-window_frame=
                    +-WindowFrame(frame_unit=ROWS)
                      +-start_expr=
                      | +-WindowFrameExpr(boundary_type=UNBOUNDED PRECEDING)
                      +-end_expr=
                        +-WindowFrameExpr(boundary_type=UNBOUNDED FOLLOWING)

    FieldList: [#0: $col1 BIGINT]
```

If you walk up the stack one level, you'll see that this function call is `retrieveRexNode(zetaNode, input.getRowType().getFieldList())` This input does not match what we expect, the input FieldList has only 1 column but the ZetaSQL input AnalyticScan has 4 columns. That input is produced in `AnalyticScanConverter`, which is the code you are adding in this PR. It looks like the ZetaSQL `AnalyticScan` passes through all of the columns from the input but those are missing from your conversion. See the description of what it does here: https://github.com/google/zetasql/blob/1933acefe8eb10e293309a9fa7d7c7f1cb6bf8c4/zetasql/analyzer/analytic_function_resolver.cc#L870

roger-mike commented:
Hi @apilloud. I added the input's columns to the ```projects``` and ```fieldNames``` of the ```AnalyticScanConverter``` result. So far, it generates a BeamPlan like this:
```
INFO: BEAMPlan>
BeamZetaSqlCalcRel(expr#0..4=[{inputs}], expr#5=[2:BIGINT], expr#6=[3:BIGINT], expr#7=[SUM($t2) OVER ()], $col1=[$t7])
  BeamIOSourceRel(table=[[beam, aggregate_test_table]])
```

Right now, there's an issue related to its execution. I'm trying to apply that query to a PCollection like [this](https://github.com/apache/beam/pull/16911/files#diff-29bb834a322689e800f66763aca5ef01fdf27102e0aed38f0fa31be474309de7R3576-R3584):
```java
    String sql = "select sum(f_int_1) over (ROWS BETWEEN 2 PRECEDING AND 3 FOLLOWING) From aggregate_test_table";

    ZetaSQLQueryPlanner zetaSQLQueryPlanner = new ZetaSQLQueryPlanner(config);
    BeamRelNode beamRelNode = zetaSQLQueryPlanner.convertToBeamRel(sql);
    PCollection<Row> stream = BeamSqlRelUtils.toPCollection(pipeline, beamRelNode);
```

but I'm getting the following error:
```
program
java.lang.NullPointerException: program
	at java.base/java.util.Objects.requireNonNull(Objects.java:246)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:612)
	at org.apache.beam.sdk.extensions.sql.meta.provider.bigquery.BeamSqlUnparseContext.toSql(BeamSqlUnparseContext.java:136)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.createSqlWindowBound(SqlImplementor.java:1049)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:983)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:747)
	at org.apache.beam.sdk.extensions.sql.meta.provider.bigquery.BeamSqlUnparseContext.toSql(BeamSqlUnparseContext.java:136)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:612)
	at org.apache.beam.sdk.extensions.sql.meta.provider.bigquery.BeamSqlUnparseContext.toSql(BeamSqlUnparseContext.java:136)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:1069)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.callToSql(SqlImplementor.java:776)
	at org.apache.beam.vendor.calcite.v1_28_0.org.apache.calcite.rel.rel2sql.SqlImplementor$Context.toSql(SqlImplementor.java:750)
	at org.apache.beam.sdk.extensions.sql.meta.provider.bigquery.BeamSqlUnparseContext.toSql(BeamSqlUnparseContext.java:136)
	at org.apache.beam.sdk.extensions.sql.zetasql.BeamZetaSqlCalcRel$Transform.expand(BeamZetaSqlCalcRel.java:142)
	at org.apache.beam.sdk.extensions.sql.zetasql.BeamZetaSqlCalcRel$Transform.expand(BeamZetaSqlCalcRel.java:117)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:548)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:499)
	at org.apache.beam.sdk.extensions.sql.impl.rel.BeamSqlRelUtils.toPCollection(BeamSqlRelUtils.java:106)
	at org.apache.beam.sdk.extensions.sql.impl.rel.BeamSqlRelUtils.toPCollection(BeamSqlRelUtils.java:48)
	at org.apache.beam.sdk.extensions.sql.zetasql.ZetaSqlDialectSpecTest.testAnalyticOver(ZetaSqlDialectSpecTest.java:3581)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.apache.beam.sdk.testing.TestPipeline$1.evaluate(TestPipeline.java:323)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

Do you know something about this error? Could you give me some pointers on this? Thank you.

roger-mike commented:
Closed due change in requirements/scope, added description of the implementation to https://docs.google.com/document/d/14Yi4oEMzqS3n9-LfSNi6Q6kQpEP3gWTHzX0HxqUksdc/edit#
